histogram gui: radio buttons on the left for mode switching

Mark-64 opened the pull request:
Partially fixes #13338 
Improve usability of histogram gui, by

1) putting a fixed group of 4 radio buttons on the left side, to quickly switch between histogram, waveform, rbg parade and vectorscope. We get rid of the annoying button "next mode"
2) putting the remaining buttons on the right side. Those are options which are consequence of the chosen main mode and therefore it is a variable group: linear/log switch, orientation, vectorscope variant, color harmonies. They are only shown if relevant
3) introduced actions for most of the above, to be able to define shortcuts (@dterrahe please make a review)

![image](https://user-images.githubusercontent.com/43290988/214552565-241dac4c-6075-4a39-9271-cd1a8c11c8f8.png)

On a side note, I decided not to introduce a new dt radio button subclass, because the implications are significant and the advantage is small.

Next steps (with separate PRs)

1) a submenu for choosing the color harmony ?
2) a switch for graph brightness (normal / high) ?
3) better icons, following the suggestions received

@Nilvus I am afraid this requires a css adaptation, see the screenshot. I will propose it with a separate commit in this PR

dterrahe commented on src/libs/histogram.c:
```diff
@@ -2173,24 +2200,19 @@ void gui_init(dt_lib_module_t *self)
+    d->scope_type_button[i] = dtgtk_togglebutton_new(dt_lib_histogram_scope_type_icons[i], CPF_NONE, NULL);
+    text = g_strdup_printf("%s %s", _("set mode to"), _(dt_lib_histogram_scope_type_names[i]));
+    gtk_widget_set_tooltip_text(d->scope_type_button[i], text);
+    dt_action_define(dark, NULL, text, d->scope_type_button[i], &dt_action_def_button);
```
Just using `dt_lib_histogram_scope_type_names[i]` as the action name would also solve the problem that `dt_action_define` needs the _untranslated_ name (so the internal identifier doesn't break when switching languages).

Mark-64 replied:
done

dterrahe reviewed (approved):
Just a few comments. 

Currently shortcuts directly to the buttons don't work because normally "hidden" buttons are "inactive" buttons. To disable this logic, the button groups need to be packed into another intermediate `GtkBox` before being put into the overlay. Then _this_ box can be (un)hidden on enter/leave rather than the left and rgb box directly. I'd be happy to make these changes (and other _action_ related changes I mentioned) after this gets merged if you like.

I hope it is clear that I feel your improvements are so good that it is worth dotting the last i's to make it "perfect".

Separately from this PR, when rotating harmonies, it would be nice to do a "mod 360" so you can "scroll through" the zero, instead of having to go almost all the way back.

dterrahe commented on src/libs/histogram.c:
```diff
@@ -2202,39 +2224,46 @@ void gui_init(dt_lib_module_t *self)
+  d->blue_channel_button = dtgtk_togglebutton_new(dtgtk_cairo_paint_color, CPF_NONE, NULL);
+  gtk_widget_set_name(d->blue_channel_button, "blue-channel-button");
+  gtk_widget_set_tooltip_text(d->blue_channel_button, d->blue ? _("hide blue channel") : _("show blue channel"));
+  gtk_toggle_button_set_active(GTK_TOGGLE_BUTTON(d->blue_channel_button), d->blue);
```
`  dt_action_define(dark, N_("toggle colors"), N_("blue"), d->blue_channel_button, &dt_action_def_toggle);`

Mark-64 replied:
done

Mark-64 commented:
@dterrahe Thank you very much for the review. I will try to implement your proposals

TurboGit commented:
> @dterrahe Thank you very much for the review. I will try to implement your proposals

@Mark-64 : You're going to do that in a separate PR, right? If so I'll review and see if this can be merged. I agree that from UI point of view this is a very nice improvement.

Mark-64 commented:
> > @dterrahe Thank you very much for the review. I will try to implement your proposals
> 
> @Mark-64 : You're going to do that in a separate PR, right? If so I'll review and see if this can be merged. I agree that from UI point of view this is a very nice improvement.

@TurboGit Actually I would like to do as much as possible here, some of the changes requested are easy to do.
I will convert to draft

Nilvus commented:
> @Nilvus I am afraid this requires a css adaptation, see the screenshot. I will propose it with a separate commit in this PR

If it's about borders missing on buttons on top left, I suggest to update that directly on that PR. Missing borders are because those buttons become toggle. On histogram in master, only RGB buttons are toggle and we don't want them to have borders or background-color override them. And as .toggle class were already, I use it for that. I suggest so to create a css function with something like `rgbtoggle` name and replace `.toggle` class by that new one on lines 670 and 1682 that start with `#main-histogram .dt_module_btn`. That should do the trick

Mark-64 commented:
@Nilvus Thanks for the suggestions, I didn't know of that function.
Unfortunately there is another issue. This
```
#main-histogram .dt_module_btn:not(.rgb_toggle)
{
  color: alpha(@field_selected_fg, 0.56);
  background-color: alpha(darker(@button_bg),0.8);
  border: 0.07em solid alpha(@button_bg, 0.33);
}
```
is now also applied to the toggle buttons on the left and they do not take anymore the `checked` state background color.
What to do?

Mark-64 commented on src/libs/histogram.c:
```diff
@@ -2202,39 +2224,46 @@ void gui_init(dt_lib_module_t *self)
+  d->scope_view_button = dtgtk_button_new(dtgtk_cairo_paint_empty, CPF_NONE, NULL);
+  ac = dt_action_define(dark, NULL, N_("switch histogram type"), d->scope_view_button, NULL);
+  dt_action_register(ac, NULL, _lib_histogram_change_type_callback, 0, 0);
+  gtk_box_pack_end(GTK_BOX(d->button_box_rgb), d->scope_view_button, FALSE, FALSE, 0);
```
done

Mark-64 commented on src/libs/histogram.c:
```diff
@@ -2202,39 +2224,46 @@ void gui_init(dt_lib_module_t *self)
-  // these are toggle boxes with a meaningful active state, unlike the type/view buttons
-  d->red_channel_button = dtgtk_togglebutton_new(dtgtk_cairo_paint_color, CPF_NONE, NULL);
-  gtk_widget_set_name(d->red_channel_button, "red-channel-button");
-  gtk_widget_set_tooltip_text(d->red_channel_button, d->red ? _("hide red channel") : _("show red channel"));
```
done

Mark-64 commented on src/libs/histogram.c:
```diff
@@ -2154,6 +2180,7 @@ void gui_init(dt_lib_module_t *self)
   dt_action_t *ac = NULL;
 
   dt_action_register(dark, N_("cycle histogram modes"), _lib_histogram_cycle_mode_callback, 0, 0);
+  dt_action_register(dark, N_("switch histogram mode"), _lib_histogram_change_mode_callback, 0, 0);
```
removed

Mark-64 commented on src/libs/histogram.c:
```diff
@@ -2268,8 +2297,9 @@ void gui_init(dt_lib_module_t *self)
 
   /* connect callbacks */
+  for(int i=0; i<DT_LIB_HISTOGRAM_SCOPE_N; i++)
+    g_signal_connect(G_OBJECT(d->scope_type_button[i]), "button-press-event", G_CALLBACK(_scope_histogram_mode_clicked), d);
```
done

Mark-64 commented on src/libs/histogram.c:
```diff
@@ -2190,15 +2165,20 @@ void gui_init(dt_lib_module_t *self)
-  gtk_button_box_set_layout(GTK_BUTTON_BOX(d->button_box_rgb), GTK_BUTTONBOX_EXPAND);
-  gtk_widget_set_valign(d->button_box_rgb, GTK_ALIGN_START);
+  d->button_box_main = gtk_box_new(GTK_ORIENTATION_HORIZONTAL, 0);
+  gtk_style_context_add_class(gtk_widget_get_style_context(d->button_box_main),"button_box");
```
done

Mark-64 commented on src/libs/histogram.c:
```diff
@@ -2190,15 +2165,20 @@ void gui_init(dt_lib_module_t *self)
+  gtk_widget_set_halign(d->button_box_main, GTK_ALIGN_START);
+
+  d->button_box_opt = gtk_box_new(GTK_ORIENTATION_HORIZONTAL, 0);
+  gtk_style_context_add_class(gtk_widget_get_style_context(d->button_box_opt),"button_box");
```
done

Mark-64 commented on src/libs/histogram.c:
```diff
@@ -2219,46 +2200,49 @@ void gui_init(dt_lib_module_t *self)
 
   // red/green/blue channel on/off
   d->blue_channel_button = dtgtk_togglebutton_new(dtgtk_cairo_paint_color, CPF_NONE, NULL);
+  gtk_style_context_add_class(gtk_widget_get_style_context(d->blue_channel_button),"rgb_toggle");
```
done

Mark-64 commented on src/libs/histogram.c:
```diff
@@ -2219,46 +2200,49 @@ void gui_init(dt_lib_module_t *self)
   gtk_box_pack_end(GTK_BOX(d->button_box_rgb), d->green_channel_button, FALSE, FALSE, 0);
 
   d->red_channel_button = dtgtk_togglebutton_new(dtgtk_cairo_paint_color, CPF_NONE, NULL);
+  gtk_style_context_add_class(gtk_widget_get_style_context(d->red_channel_button),"rgb_toggle");
```
done

Mark-64 commented:
I have implemented all the requested changes.
PR rebased, squashed in two commits and ready to go.
All the actions work even if buttons are hidden.

Nevertheless, I appreciate a further review from @dterrahe and @Nilvus 
Thanks

Nilvus commented:
Just really great UI improvement. This will maybe make me use again histogram. Access to modes is faster and we better see them. Great work and thanks a lot @Mark-64.

TurboGit reviewed (approved):
Nice work, really, thanks!

TurboGit commented on src/libs/histogram.c:
```diff
@@ -2146,79 +2135,100 @@ void gui_init(dt_lib_module_t *self)
+  for(int i=0; i<DT_LIB_HISTOGRAM_SCOPE_N; i++)
+  {
+    d->scope_type_button[i] = dtgtk_togglebutton_new(dt_lib_histogram_scope_type_icons[i], CPF_NONE, NULL);
+    text = g_strdup_printf("%s %s", _("set mode to"), _(dt_lib_histogram_scope_type_names[i]));
```
@Mark-64 : After translation I find this **"set mode to"** superfluous. Just the name of the scope when flying over the icon seems enough. Ok to remove?

Mark-64 replied:
ok

Mark-64 commented:
RELEASE NOTES:
The histogram gui has been reworked. Control buttons have been splitted in two groups: on the left side, a series of buttons to switch among histogram modes (histogram, waveform, rbg parade, vectorscope). On the right side, the buttons that control the aspect of each mode (rgb channels, orientation, vectorscope type). For the RYB vectorscope, a series of buttons have been added to visualize guide lines for the most common color harmonies.

TurboGit commented:
Thanks!
